### DISCUSSION TITLE
would like for device_OnPacketArrival to be async, but PacketCapture e doesn't support

### DISCUSSION THREAD

**Discussion by TXAE** [chmorgan/sharppcap#399] [General]
Hi,

I would like for device_OnPacketArrival to be async, but PacketCapture e doesn't support that.
It is an OnPacketArrival void:
`device.OnPacketArrival += new PacketArrivalEventHandler(device_OnPacketArrival);`
Is this planned or am I doing somethin wrong?

`private static async void device_OnPacketArrival(object sender, PacketCapture e) { /* magic */ }`
==> Gives me CS4012: Variables of type PacketCapture cannot be declared in async methods

**Comment by kayoub5** [comment]
You need to use `e.GetPacket()` to get the `RawCapture` object, then you can use that in an async method.

**Reply by TXAE** [reply]
For a PacketArrivalEventHandler-Method, the use of `object sender, PacketCapture e` as parameters is a must. I cannot declare the method async when obligatory parameters cannot be used in an async method, no matter how I obtain the RawCapture object in that method.

![image](https://user-images.githubusercontent.com/70020564/172054582-dde3dac8-2f50-43fd-9379-d742580da88b.png)

**Reply by kayoub5** [reply]
You need a wrapper
```cs
        private static void device_OnPacketArrival(object sender, PacketCapture e)
        {
            device_OnPacketArrivalAsync(sender, e.GetPacket());
        }
        private static async void device_OnPacketArrivalAsync(object sender, RawCapture e)
        {
            // Logic here
        }
```

**Reply by TXAE** [reply]
it worked! :) Thank you!
Use case: https://github.com/TXAE/CIROS-file-to-MQTT-Publisher
